### ISSUE TITLE
NixOS 18.03 minimal fails fresh install (systemd-boot-builder)

### ISSUE THREAD

**Issue by slimey** [NixOS/nixpkgs#38477]
## Issue description

Installing Nixos from the 18.03 minimal installation DVD iso, for a fresh 64-bit Hyper-V vm target system, is failing for me at `Failed to create EFI Boot variable entry: No such file or directory`:
```
# nixos-install
building the configuration in /mnt/etc/nixos/configuration.nix...
[168 built, 274 copied (1188.0MiB), 23.9 MiB DL]
copying channel...
installing the boot loader...
setting up /etc...
Initializing machine ID from random generator.
Created "/boot/EFI".
Created "/boot/EFI/systemd".
Created "/boot/EFI/BOOT".
Created "/boot/loader".
Created "/boot/loader/entries".
Copied "/nix/store/dzrd8yxh9r1bwisphg3z47dxrxlgskv1-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/dzrd8yxh9r1bwisphg3z47dxrxlgskv1-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
Failed to create EFI Boot variable entry: No such file or directory
Traceback (most recent call last):
  File "/nix/store/71qjm4rjlxzz7z4dsxd8dndj9dlyil38-systemd-boot-builder.py", line 210, in <module>
    main()
  File "/nix/store/71qjm4rjlxzz7z4dsxd8dndj9dlyil38-systemd-boot-builder.py", line 185, in main
    subprocess.check_call(["/nix/store/dzrd8yxh9r1bwisphg3z47dxrxlgskv1-systemd-237/bin/bootctl", "--path=/boot", "install"])
  File "/nix/store/lp7x6ziq5b2zihgad0i8a28xxvz5vnrr-python3-3.6.4/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/nix/store/dzrd8yxh9r1bwisphg3z47dxrxlgskv1-systemd-237/bin/bootctl', '--path=/boot', 'install']' returned non-zero exit status 1.
```

Running under strace, there is a *possible* match at
```
openat(AT_FDCWD, "/sys/firmware/efi/efivars/Boot0000-8be4df61-93ca-11d2-aa0d-00e098032b8c", O_WRONLY|O_CREAT|O_NOCTTY|O_CLOEXEC, 0644) = -1 ENOENT (No such file or directory)
```
but this was from a quick grep from a lot of content, so might well be a red herring. That path is present in the host (CD-booted) system. Full trace [here](http://lpaste.net/364449).

### Steps to reproduce

I tried to follow the vanilla NixOS installation instructions as closely as possible.

## Technical details

[This is the installation CD system]

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.32, NixOS, 18.03.131792.becbe4dbe16 (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0`
 - channels(root): `"nixos-18.03.131792.becbe4dbe16"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

In case relevant, the hardware-configuration.nix contents are:
```
# Do not modify this file!  It was generated by ‘nixos-generate-config’
# and may be overwritten by future invocations.  Please make changes
# to /etc/nixos/configuration.nix instead.
{ config, lib, pkgs, ... }:

{
  imports = [ ];

  boot.initrd.availableKernelModules = [ "sd_mod" "sr_mod" "hv_storvsc" ];
  boot.kernelModules = [ ];
  boot.extraModulePackages = [ ];

  fileSystems."/" =
    { device = "/dev/disk/by-uuid/992d6869-7bf5-4c4d-a25b-5eca354b731c";
      fsType = "ext4";
    };

  fileSystems."/boot" =
    { device = "/dev/disk/by-uuid/57D8-AB48";
      fsType = "vfat";
    };

  swapDevices =
    [ { device = "/dev/disk/by-uuid/dbdbddff-d580-42dd-b553-b205b2e56ff9"; }
    ];

  nix.maxJobs = lib.mkDefault 4;
  services.xserver.videoDrivers = [ "fbdev" ];
}
```

**Comment by bcarrell** [comment]
Having the same issue right now.

**Comment by anselmio** [comment]
Getting the same using `nixos-graphical-18.03.131800.ac3437aa061-x86_64-linux.iso` for a fresh install attempt. Identical installation steps with `nixos-graphical-17.09.3205.12512f25e9e-x86_64-linux.iso` succeeds.

**Comment by bcarrell** [comment]
Confirmed that `nixos-minimal-17.09.3205.12512f25e9e-x86_64-linux.iso` also succeeds in my case.

---

- `nixos-minimal-17.09.3205.12512f25e9e-x86_64-linux.iso` :heavy_check_mark: 
- `nixos-minimal-18.09pre133640.ea145b68a01-x86_64-linux.iso` :x: 
- `nixos-minimal-18.03.131802.4b4bbce199d-x86_64-linux.iso` :x:

**Comment by Redcroft** [comment]
I'm having the same issue on an Acer ES1-132

**Comment by smashedtoatoms** [comment]
Yep, here too on an XPS 13 9370

**Comment by LSLeary** [comment]
I had the same issue. My workaround was to temporarily set
```
boot.loader.efi.canTouchEfiVariables = false;
```
after which the install would succeed but not write the boot-listing efivar. Then I manually created that efivar with efibootmgr. I imagine the command would have worked from the installation media, but I'm not sure since I ran it in Debian.
```
efibootmgr -c -d /dev/sdb -p 9 -L "SystemD" -l "\efi\systemd\systemd-bootx64.efi"
```
(on my system the EFI boot partition is /dev/sdb9)

There were also *.efi files generated in \efi\nixos, but I didn't try any since it booted with this.

**Comment by andrewchambers** [comment]
is anyone actively working on this? what does severity blocker signify?

**Comment by adamtulinius** [comment]
Also seeing this when installing on vmware, but with custom install media built from the nixos-18.03channel as of today (c0c5571ec1a33a5109c122b8821eb15c36cbb1ec). No problems with 17.09. EFI is certainly enabled on the host.

**Comment by BenKerry** [comment]
I run into the exact same problem while trying to install NixOS 18.03 into VMware.

**Comment by nenadstojanovikj** [comment]
I got the same behaviour when I tried to install NixOS 18.03 on Dell XPS 13 (9360).

**Comment by forgemo** [comment]
Same issue when tying a fresh install of nixos-minimal-18.03.132021.c0c5571ec1a-x86_64-linux.iso on VirtualBox.  "EFI" option in virtual machine system settings is enabled. I followed the install instructions as closely as possible.

**Comment by basvandijk** [comment]
I got the same issue when installing NixOS-18.03 (ad222f3c1433e77b6d8483fd93ea7e7357ad0424) on a recent Intel NUC.

**Comment by samueldr** [comment]
A fix has been pushed to the master and 18.03 branches. It will take some time to reach the channel, and for a new ISO to be generated using the fix, but this specific issue should be fixed soon for new installs.

In the meantime, [this workaround](https://github.com/NixOS/nixpkgs/issues/38477#issuecomment-379417134) should work. Once installed, the `canTouchEfiVariables` option can be put back to the defaults (removed from your configuration) as the issue only affects the installation environment.

**Comment by joseph-long** [comment]
For the record, I was unable to boot using the aforelinked efibootmgr workaround on a 2018 Lenovo X1 Yoga. I had a pre-existing Linux boot entry, which I removed, but the new entry that I added did not appear as a startup option.

**Comment by slimey** [comment]
Confirmed (what Wolfereign said) - thanks!

**Comment by andrewchambers** [comment]
I think this issue was automatically closed by the github merge bot, but shouldn't have been.

**Comment by samueldr** [comment]
4 days ago, it shouldn't have been (as github doesn't know about nixpkgs channels).

The current iso image on nixos.org is made from commit `b50443b5c4a`, which is after `e932b5be26`, and [includes it](https://github.com/NixOS/nixpkgs-channels/commit/e932b5be269f6264860ae7004de66a85f42fe3a7).

Simply put, the 18.03 iso *do contain the needed fix for this specific issue*. The iso image on the website was updated somewhere between ~24h ago (when the channel advanced) and a couple hours ago when I checked whether the iso image changed. I have also, just in case, verified that it does not reproduce the fixed issue.

I'm using this post to add this note:

> **If you have a fresh image from nixos.org and it fails with this issue or a similar one, please open an issue and/or [contact the community](https://nixos.org/nixos/support.html).**
>
> This specific issue *should be fixed*.
